Fixed the error for kubectl edit multiple resources

janetkuo opened the pull request:
Fixes #15907

2 bugs are fixed:
1) panic when editing a list object
2) editing a list of objects failed to update (without an error)

``` console
$ kubectl get pods 
NAME         READY     STATUS    RESTARTS   AGE
mock-9e7dn   1/1       Running   0          3m
mock-adkeu   1/1       Running   0          3m

$ kubectl edit pods 
# Do some edit, and then save...
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x73edf2]

goroutine 1 [running]:
k8s.io/kubernetes/pkg/api/meta.(*genericAccessor).SetAnnotations(0xc208031cc0, 0xc20818f7d0)
    <autogenerated>:20 +0xd2
k8s.io/kubernetes/pkg/kubectl.GetModifiedConfiguration(0xc208221e80, 0xc208280200, 0x0, 0x0, 0x0, 0x0, 0x0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/apply.go:102 +0x224
k8s.io/kubernetes/pkg/kubectl.UpdateApplyAnnotation(0xc208221e80, 0x0, 0x0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/apply.go:169 +0x3f
k8s.io/kubernetes/pkg/kubectl/cmd.RunEdit(0xc2080b6f00, 0x7f05615e6fb0, 0xc20802e008, 0xc20813de00, 0xc208190010, 0x1, 0x1, 0x1bda6d0, 0x0, 0x0, ...)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/edit.go:214 +0x1609
k8s.io/kubernetes/pkg/kubectl/cmd.func·018(0xc20813de00, 0xc208190010, 0x1, 0x1)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/edit.go:83 +0x7e
```

cc @bgrant0607 @jackgr @kubernetes/kubectl this should be v1.1

janetkuo committed:
Fixed the error for kubectl edit multiple resources

JanetKuo commented on pkg/kubectl/cmd/edit.go:
```diff
@@ -145,139 +145,142 @@ func RunEdit(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []strin
-			return preservedFile(err, file, out)
-		}
+			// parse the edited file
+			updates, err := rmap.InfoForData(edited, "edited-file")
```
When we edit a list of objects (`edited` is a `[]byte` type List object), `InfoForData` ignores the changes users made in each Item of the List, and therefore the changes are never patched when you edit a list. So instead, I parse `infos` as []runtime.Object in L148 and the users now need to traverse/edit them one by one (not editing the List directly). 
@smarterclayton Do you also see this in Openshift origin's edit?

deads2k replied:
> @smarterclayton @deads2k Do you also see this in Openshift origin's edit?

I'm been able to edit lists with `oc edit`.

JanetKuo commented:
Re-assigning to @jlowdermilk

bgrant0607 commented:
We want to cherrypick this into 1.1.

@kargakis Could you take a look?

jlowdermilk commented:
Change lgtm. It's unfortunate you have to reopen the editor for each file in a list, but it's better than it not working at all.

JanetKuo commented:
Applying lgtm label since got LGTM-ed.

kargakis commented:
Maybe not the right place to ask but I'll go ahead... How about reloading the file in case of a syntax error? Currently `kubectl edit` will exit if a user makes a syntax error. Wouldn't it be better if we returned the edited file and pointed to the user where did they mess up? In `oc edit` and more specifically https://github.com/openshift/origin/pull/2993 I have been working on reloading the edited file and it works fine .... except one thing. The error reporting the bad line (returned from the yaml package) will have the wrong line number since we are adding more sentences to the header than what it had by the time the error happened. All solutions I have tried so far are hacks (add a placeholder for the error?). Any thoughts?

kargakis commented on pkg/kubectl/cmd/edit.go:
```diff
@@ -145,139 +145,142 @@ func RunEdit(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []strin
-			} else {
-				os.Remove(file)
+			glog.V(4).Infof("User edited:\n%s", string(edited))
+			fmt.Printf("User edited:\n%s", string(edited))
```
Why are you printing the whole file by default? IMHO glog is enough.

JanetKuo replied:
Removed.

kargakis commented on pkg/kubectl/cmd/edit.go:
```diff
@@ -145,139 +145,142 @@ func RunEdit(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []strin
-				preservedFile(nil, file, out)
-			} else {
-				os.Remove(file)
+			if bytes.Equal(original, edited) {
```
Note that this compares comments too. We need something like https://github.com/kargakis/origin/blob/e65b2db16414459318a03e07de0fcd6540e5a655/pkg/cmd/cli/cmd/edit.go#L215

JanetKuo replied:
Added `stripComments`

JanetKuo commented:
@kargakis thanks. I've addressed your comments. As for exit or not when there's a syntax error, I agree that reloading the file would benefit the user. However, when the users want to edit a list of resource, the editor would reload too, and I'm afraid that'd confuse the user (am I editing the next item or did I do something wrong?). It's not as obvious as exiting directly and tells the user that there's a syntax error imo. 

We can discuss and implement this reloading feature in another PR.

kargakis commented:
If I tried to edit a list I would expect to edit all of the items at once. What would it take to make that change?

JanetKuo commented:
@kargakis as mentioned above https://github.com/kubernetes/kubernetes/pull/15980#discussion_r42564695, if the users edit a list of resource, `InfoForData` won't return the changes users made in each Item of the List, and therefore the changes are never patched. That's why this PR separate each item from the list.

kargakis commented:
Ok, I guess it's better than not working at all

This LGTM

JanetKuo commented:
Squashed commits.

JanetKuo commented:
@k8s-bot test this

janetkuo commented:
Need to cherry-pick #15010 and #15042 in order to cherry-pick this PR.
